Update modifier state on key release using key code

jjttjj opened the pull request:
Hi, I noticed some problems with [`clojure-keymap`](https://github.com/phronmophobic/clobber/blob/main/src/com/phronemophobic/clobber/modes/clojure/ui.clj#L1448).

In summary, I am pretty sure the code dealing with removing the modifier state on key release [here](https://github.com/phronmophobic/clobber/blob/913ab1206dbe02a4c019d74ade9f24a24bdbbb86/src/com/phronemophobic/clobber/modes/clojure/ui.clj#L1518-L1520) is broken for me and is causing issues with certain keybindings.

At first I thought I could just remove the `not` on those lines and that seems to work for me. However, if found that the mods bitmap will contain the mask of the mod being released as well as any that are still down, so it could incorrectly remove a still held down mod key from the state. I think we need to use the actual mod `key` in the key-event.

I could refactor this and make it more efficient if you want, so we're not checking all these sets on every key release, wanted to run it by you first though because I'm not 100% this is the correct approach and wont break anything, though it seems to work for me.

Backstory, in case it's useful:

I found that I was struggling to get the `C-c )` keybinding for `paredit-forward-slurp-sexp` working. It seemed to work roughly half the time. Still not 100% why, but ultimately I came to suspect that the minimal demonstration of the issue could be seen if you do the following key sequence:
1. Press control
2. Press and release any other (non modifier) key
3. Press and release any other (non modifier) key again
4. Release control
If you do this, `:ctrl?` will still be left in the modifier state (it wont be "applied" to the next key, it just messes up the matching process I believe).

jjttjj committed:
Update modifier state on key release using key code

phronmophobic commented:
> I found that I was struggling to get the C-c ) keybinding for paredit-forward-slurp-sexp working

I've also had this same exact issue! The key binding code definitely needs to be cleaned up. My guess is that pressing "shift" to input ")" instead of "0" is causing the keybindings state to reset.

Anyway, the current code is hacky and should be redone properly

jjttjj committed:
Attempt at simplifying clojure-keymap

I removed the :key-press handler in favor of trying to do everything with the
:key-event handler. There is still some awkardness here but at least the surface
area is smaller.

jjttjj committed:
Fix shift capitalization

jjttjj committed:
No mod keys should interrupt key chord

phronmophobic commented:
I'm adding generative tests and fixing a bunch of edge cases that were found. I'll try to review this once that is complete which will hopefully be soon!

jjttjj commented:
Sounds good! I regret making this a PR instead of an issue, it's more involved than I thought. I think I'm making progress but I'm still discovering some behaviors that might need tweaking, and there's likely some tidying up to do as well as higher level decisions that might factor in. I will continue to kick the tires on this and make tweaks here, but feel free to close this PR for now if you don't want all the pings (and I could come up an issue and a more coherent PR later if so).

phronmophobic commented:
Yea, my intent is to move the keybinding code to its own namespace. Getting it right is tricky because you sorta have to reverse engineer what membrane, glfw, and the operating system are doing. It's further complicated by the fact that `key-event` is about keyboard keys and `key-press` is about the text produced by a key (which I think can be configured by custom keyboard layouts like dvorak). To implement it correctly, you need to interpret both event types correctly.

phronmophobic commented:
Fyi, I finally got a chance to look into this. I refactored all the key-binding stuff into its own namespace. 

 https://github.com/phronmophobic/clobber/blob/main/src/com/phronemophobic/clobber/util/ui/key_binding.clj

I ended up implementing everything in terms of `key-event` instead of a combination of `key-event` and `key-press`. The problem is that if you're holding a modifier key like alt or ctrl, then `key-press` won't get called so listening to key-press doesn't help that much. After reading the [glfw input guide](https://www.glfw.org/docs/latest/input_guide.html), I'm still not totally sure if the `key` argument is the keyboard key before or after considering OS settings. There might be some issues for custom keyboard layouts, but it's no worse than before.

I also fixed the problem where `"C-c C-)"` wasn't working because pressing ctrl was reseting the key-binding state.

At some point, it's probably worth figuring out a good way to read OS keyboard settings (if it's not already done by glfw), but I'll probably prioritize other work for now unless someone complains.

jjttjj commented:
Awesome, thanks for the fix and the update!
